Refactor martha_v3 handler into smaller functions

kshakir opened the pull request.

kshakir committed:
Refactor martha_v3 handler into smaller functions

mcovarr commented on martha/martha_v3.js:
```diff
@@ -468,22 +460,45 @@ async function marthaV3Handler(req, res) {
+    Object.assign(opt, {
+        requestedFields,
+        httpsMetadataUrl,
+        auth,
```
It looks like `auth` and `requestedFields` were not changed since their creation by destructuring from `opt` (they are `const`s), so is this just restructuring the same values back into `opt`?

mcovarr commented on martha/martha_v3.js:
```diff
@@ -426,37 +427,28 @@ function overlapFields(requestedFields, serviceFields) {
-
-    res.status(200).send(partialResponse);
-}
+function parseOptions(opt) {
```
nit: some of these "options" are not optional, e.g. `url` and `auth`. Perhaps `parseParameters(params)` or something like that?

mcovarr commented on martha/martha_v3.js:
```diff
@@ -468,22 +460,45 @@ async function marthaV3Handler(req, res) {
+    });
+}
+
+async function retrieveData(opt) {
```
more non-options in this `opt`

mcovarr replied:
Unrelated to your changes but it might be helpful to have a link to Michael's Lucid Chart sequence diagram(s) here.

mcovarr commented on martha/martha_v3.js:
```diff
@@ -534,12 +546,70 @@ async function marthaV3Handler(req, res) {
+    });
+}
+
+function makeResponse(opt) {
```
here too

kshakir committed:
PR Fixup: renamed functions, removed duplicate params, added doc links

breilly2 commented on martha/martha_v3.js:
```diff
@@ -426,37 +427,28 @@ function overlapFields(requestedFields, serviceFields) {
     } catch (error) {
-        logAndSendBadRequest(res, error);
-        return;
+        throw new BadRequestError(error);
```
`validateRequest` could throw `BadRequestError`s, and then this function doesn't need to `try`/`catch`.

kshakir committed:
Merge branch 'bt_170_access_url' into ks_refactor_complexity

breilly2 commented on martha/martha_v3.js:
```diff
@@ -534,12 +552,70 @@ async function marthaV3Handler(req, res) {
+            logAndSendServerError(res, error.cause, error.description);
+        } else {
+            console.error(`Uncaught error: ${error}`);
+            throw error;
```
Since we chatted about this a little in the mob, I remembered that letting an exception escape from a cloud function causes GCP to destroy that instance instead of letting it continue handling requests. Since we don't know what went wrong by the time we get here, killing the instance is probably a good idea. 👍

breilly2 commented on martha/martha_v3.js:
```diff
@@ -468,22 +460,51 @@ async function marthaV3Handler(req, res) {
 
     let drsResponse;
     try {
         drsResponse = responseParser(response);
```
Can this even throw an error? I didn't notice anything in `responseParser`.

breilly2 replied:
Nevermind. There's code in `responseParser` that deals with converting DOS responses to DRS responses that I suspect could throw errors because I don't have enough knowledge of the DOS spec. I retract my implied suggestion to remove the `try`/`catch` from here.

breilly2 reviewed (approved):
👍   Looks good. Only very minor questions/nits.

kshakir committed:
Multiple updates:
- Don't wrap an Error in a BadRequestError, just throw a BadRequestError
- Moved some variable destructuring based on upstream changes
